fabtests/benchmarks/benchmark_shared.c

```diff
@@ -267,23 +266,43 @@ int bandwidth(void)
 	return 0;
 }
 
-static int bw_rma_comp(enum ft_rma_opcodes rma_op)
+/**
+ * @brief Get completions of RMA operations, and verify_data if requested
+ *
+ * Completions are received for the rma_op, according to the sequence counter.
+ * Verifications assume <valid_windows> many operations of opts.transfer_size
+ * were completed starting at buf + offset_rma_start.
+ *
+ * @return 0 on success.
+*/
+static int bw_rma_comp(enum ft_rma_opcodes rma_op, int valid_windows)
 {
 	int ret;
 
 	if (rma_op == FT_RMA_WRITEDATA) {
+		/* for write data, only the client sends,
+		 * and only the server verifies. */
 		if (opts.dst_addr) {
 			ret = bw_tx_comp();
+			return ret;
 		} else {
 			ret = rma_bw_rx_comp();
 		}
 	} else {
 		ret = ft_get_tx_comp(tx_seq);
+		if (ret)
+			return ret;
+
+		if (rma_op == FT_RMA_WRITE && ft_check_opts(FT_OPT_VERIFY_DATA)) {
+			ft_sync();
+		}
 	}
-	if (ret)
+	if (ret || !ft_check_opts(FT_OPT_VERIFY_DATA))
 		return ret;
 
-	return 0;
+	ret = ft_check_buf(rx_buf + offset_rma_start,
```

Comment:
The structure of the above code is confusing.  Having a check for error and returning early is acceptable.  e.g.
```
if (ret)
    return ret;
```
Note that we won't need this check if the code above is restructured.  The call to ft_check_buf() is dependent on ft_check_opts(VERIFY).  That's easier to understand when placed together.  E.g.
```
if (ft_check_opts(VERIFY))
    ret = ft_check_buf(...)

return ret;
```

Reply:
Restructured.  Hopefully it is more clear now.